Move rules_pkg_lib to //pkg

jacky8hyf opened the pull request:
... next to the .bzl files. This is because the
//doc_build package is not loadable when rules_pkg itself
is not the root Bazel module, because stardoc is a
dev_dependency. So, //doc_build:rules_pkg_lib is not
usable by clients of rules_pkg.

Move the target to //pkg, next to the .bzl files.

Also add @rules_python//python:defs_bzl used by pkg_install.

Link: https://github.com/bazelbuild/rules_pkg/issues/897

cgrindel commented:
@jacky8hyf Do you want to address the failing tests before we review?

jacky8hyf commented:
> @jacky8hyf Do you want to address the failing tests before we review?

I don't think the test error sare related to my change. The errors are: 

```
 ERROR: no such package '@@[unknown repo 'mappings_test_external_repo' requested from @@]//pkg': The repository '@@[unknown repo 'mappings_test_external_repo' requested from @@]' could not be resolved: No repository visible as '@mappings_test_external_repo' from main repository. Was the repository introduced in WORKSPACE? The WORKSPACE file is disabled by default in Bazel 8 (late 2024) and will be removed in Bazel 9 (late 2025), please migrate to Bzlmod. See https://bazel.build/external/migration.
```

`mappings_test_external_repo` is defined in WORKSPACE file. Perhaps `--noenable_workspace` is set because it is using Bazel 8: (8.0.0-pre.20240911.1 from the test).

I can't reproduce the same error offline.

cgrindel commented:
@aiuto Any thoughts on the test failures?

jacky8hyf commented:
gentle ping

aiuto commented:
I do think the failing tests are unrelated, but I am dubious about the change.
We are trying to limit taking on new dependencies. Skylib depends on rules_pkg, so the mutual embrace makes compatibility_level updates really hard.

jacky8hyf commented:
AFAIK there are no **new** dependencies. See this line:

https://github.com/bazelbuild/rules_pkg/blob/5c6aec64627c91f8ae71820173766237311d4ac6/pkg/BUILD#L15

This means for anyone that relies on the `@rules_pkg//pkg` package, the `starlark_library.bzl` is loaded, which relies on skylib because of

https://github.com/bazelbuild/rules_pkg/blob/5c6aec64627c91f8ae71820173766237311d4ac6/pkg/private/make_starlark_library.bzl#L6

In addition, skylib is already a **non-dev** dependency:

https://github.com/bazelbuild/rules_pkg/blob/5c6aec64627c91f8ae71820173766237311d4ac6/MODULE.bazel#L11

(Skylib does depend on rules_pkg as a dev-dependency, on the other hand: 
https://github.com/bazelbuild/bazel-skylib/blob/e8e9d218ff563f93e3a7f7547ea532e66b210620/MODULE.bazel#L20
)

I am not changing MODULE.bazel for rules_pkg so there are no dependency changes. Am I missing something?

jacky8hyf commented on pkg/private/make_starlark_library.bzl:
```diff
@@ -9,13 +9,10 @@ def _make_starlark_library(ctx):
     direct = []
     transitive = []
     for src in ctx.attr.srcs:
-        if StarlarkLibraryInfo in src:
```
No; in fact I can drop this commit from this particular pull request.

I just find that it is incorrect here because 

```
transitive.append(src[StarlarkLibraryInfo])
```

This is appending the `StarlarkLibraryInfo`, not the file, to the transitive list, causing the depset creation below to fail when there is a bzl_library/starlark_library in `srcs`. I found this bug when I was trying to add rules_python/skylib to `//pkg:bzl_srcs` directly. But later I didn't actually add rules_python/skylib to `//pkg:bzl_srcs`.

Do you want me to drop this commit from this pull request and create a separate pull request?

jacky8hyf replied:
In other words, if I were to use `starlark_library` instead of `bzl_library` for the newly added `//pkg:rules_pkg_lib` target, then I do need this patch (or even fix it properly). Though, I think `@bazel_skylib//lib:paths` and `@rules_python//python:defs_bzl` are both `bzl_library`s, and due to this:

https://github.com/bazelbuild/bazel-skylib/blob/56b235e700ddd6a15b7d9fa1803fa7a84048471e/rules/private/bzl_library.bzl#L37

It is actually okay to just use the default files from these libraries without poking into StarlarkLibraryInfo, so this branch could potentially be deleted. 

Though, a pedantic person may also argue that, for this to be 100% correct, we should use the fields in StarlarkLibraryInfo from dependencies to construct StarlarkLibraryInfo of this starlark_library...

aiuto replied:
Based on https://github.com/bazelbuild/bazel/issues/18391#issuecomment-2445364390

it seems we can drop the use of starlarklibrary and bzl_library.

It seem to work for me by changing pkg/BUILD:bzl_srcs to a filegroup(name ="rules_pkg_lib"
and referring to that in doc_build/BUILD, deleting the existing rules_pkg_lib there.

jacky8hyf replied:
> changing pkg/BUILD:bzl_srcs to a filegroup(name ="rules_pkg_lib"

bzl_srcs has visibility public. Wouldn't changing it to a filegroup breaks an API?

I can push a change that does this.

jacky8hyf replied:
Also, as I said in the original commit message:

> This is because the
//doc_build package is not loadable when rules_pkg itself
is not the root Bazel module, because stardoc is a
dev_dependency.

So how should users of rules_pkg build stardoc for a `.bzl` that relies on a .bzl from rules_pkg? It seems to me that the `bzl_library` must be outside of `//docs_build`. I choose `//pkg` because it is next to the .bzl files, which is what the other projects usually do as well, e.g. rules_cc does this:

https://github.com/bazelbuild/rules_cc/blob/34f0e1f038cff9bf28d65d101c806b48f35bf92e/cc/BUILD#L85

https://github.com/bazelbuild/rules_cc/blob/34f0e1f038cff9bf28d65d101c806b48f35bf92e/docs/BUILD#L30

jacky8hyf committed:
docs: //pkg:bzl_srcs update deps.

This change updates deps for //pkg:bzl_srcs to include paths and
rules_python, which comes from the individual rules.

This bzl_library is next to the .bzl files. Users of rules_pkg may build
stardoc for their own extensions by relying on this bzl_library
directly.

This change deletes the unnecessary starlark_library rule, because
starlark_library itself already creates the non-dev dependency from
rules_pkg to skylib. A user of rules_pkg will have to have skylib
anyways. So let's just use skylib's bzl_library directly. It is okay
to put non .bzl files in bzl_library.

Test: locally update latest.md, no significant difference other than
  small changes due to the file being outdated in the first place.
Link: https://github.com/bazelbuild/rules_pkg/issues/897

jacky8hyf commented:
> I do think the failing tests are unrelated, but I am dubious about the change. We are trying to limit taking on new dependencies. Skylib depends on rules_pkg, so the mutual embrace makes compatibility_level updates really hard.

(combining with your comment in https://github.com/bazelbuild/rules_pkg/issues/897#issuecomment-2409506087 )

If I understand correctly, are you planning to make `rules_pkg` depend on `skylib` with a `dev_dependency = True` in the future? Right now it isn't, and this change is not making the problem worse.

jacky8hyf commented:
Actually, I don't think it would be possible for rules_pkg to depend on skylib as a dev_dependency, because of this usage here:

https://github.com/bazelbuild/rules_pkg/blob/5c6aec64627c91f8ae71820173766237311d4ac6/pkg/mappings.bzl#L30

... unless you planning to fork your own `paths` implementation in rules_pkg...

jacky8hyf commented:
ping, any updates?
